nemo/collections/tts/helpers/common.py

```diff
@@ -0,0 +1,956 @@
+# Copyright (c) 2022, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+###############################################################################
+
+import ast
+from typing import Tuple
+
+import numpy as np
+import torch
+from torch import nn
+from torch.cuda import amp
+from torch.cuda.amp import autocast as autocast
+from torch.nn import functional as F
+
+from nemo.collections.tts.helpers.partialconv1d import PartialConv1d as pconv1d
+from nemo.collections.tts.helpers.splines import (
+    piecewise_linear_inverse_transform,
+    piecewise_linear_transform,
+    unbounded_piecewise_quadratic_transform,
+)
+
+
+def update_params(config, params):
+    for param in params:
+        print(param)
+        k, v = param.split("=")
+        try:
+            v = ast.literal_eval(v)
+        except:
+            pass
+
+        k_split = k.split('.')
+        if len(k_split) > 1:
+            parent_k = k_split[0]
+            cur_param = ['.'.join(k_split[1:]) + "=" + str(v)]
+            update_params(config[parent_k], cur_param)
+        elif k in config and len(k_split) == 1:
+            print(f"overriding {k} with {v}")
+            config[k] = v
+        else:
+            print("{}, {} params not updated".format(k, v))
+
+
+def get_mask_from_lengths(lengths):
+    """Constructs binary mask from a 1D torch tensor of input lengths
+
+    Args:
+        lengths (torch.tensor): 1D tensor
+    Returns:
+        mask (torch.tensor): num_sequences x max_length x 1 binary tensor
+    """
+    max_len = torch.max(lengths).item()
+    ids = torch.arange(0, max_len, out=torch.cuda.LongTensor(max_len))
+    mask = (ids < lengths.unsqueeze(1)).bool()
+    return mask
+
+
+@torch.jit.script
+def fused_add_tanh_sigmoid_multiply(input_a, input_b):
+    t_act = torch.tanh(input_a)
+    s_act = torch.sigmoid(input_b)
+    acts = t_act * s_act
+    return acts
+
+
+class ExponentialClass(torch.nn.Module):
```

Comment:
Please move all `torch.nn.models` out of helpers and into the modules folder. For simple models like ConvNorm, LengthRegulator, etc. Can you search NeMo and see if similar modules already exist instead of redefining it?

Reply:
Done